File: api/sources/translate.py
"""Google Translate unofficial endpoint — EN→VI with persistent per-item cache."""
from __future__ import annotations
import asyncio
import urllib.parse
from .http import client

TRANSLATE_URL = "https://translate.googleapis.com/translate_a/single"

# id -> {title_vi, summary_vi}. Survives process lifetime; lost on restart but
# we re-translate quickly and Google Translate caches identical inputs anyway.
_trans_cache: dict[str, dict[str, str]] = {}
_lock = asyncio.Lock()


async def _translate_one(text: str) -> str:
    """Translate a single English string to Vietnamese. Returns original on failure."""
    if not text:
        return text
    text = text[:5000]  # endpoint limit
    params = {
        "client": "gtx",
        "sl": "en",
        "tl": "vi",
        "dt": "t",
        "q": text,
    }
    try:
        r = await client().get(TRANSLATE_URL, params=params, timeout=15.0)
        r.raise_for_status()
        data = r.json()
        # Response shape: [[ [translated, original, ...], [translated2, ...], ... ], ...]
        if not isinstance(data, list) or not data or not data[0]:
            return text
        parts = [seg[0] for seg in data[0] if isinstance(seg, list) and seg and isinstance(seg[0], str)]
        return "".join(parts).strip() or text
    except Exception:
        return text


async def translate_items(items: list[dict]) -> list[dict]:
    """Add `title_vi` and `summary_vi` to each item, cached by item id."""
    # Find items needing translation
    todo: list[tuple[int, str, str]] = []  # (index, title, summary)
    for i, it in enumerate(items):
        cached = _trans_cache.get(it["id"])
        if cached:
            it["title_vi"] = cached.get("title_vi") or it.get("title", "")
            it["summary_vi"] = cached.get("summary_vi") or it.get("summary", "")
        else:
            todo.append((i, it.get("title") or "", it.get("summary") or ""))

    if not todo:
        return items

    async def _do(idx: int, title: str, summary: str):
        # Limit concurrency: translate title + summary sequentially per item
        t_vi = await _translate_one(title)
        s_vi = await _translate_one(summary) if summary else ""
        return idx, t_vi, s_vi

    # Batch with limited concurrency to avoid rate-limit
    sem = asyncio.Semaphore(6)

    async def _bounded(idx: int, t: str, s: str):
        async with sem:
            return await _do(idx, t, s)

    results = await asyncio.gather(*(_bounded(i, t, s) for i, t, s in todo))

    async with _lock:
        for idx, t_vi, s_vi in results:
            item = items[idx]
            item["title_vi"] = t_vi
            item["summary_vi"] = s_vi
            _trans_cache[item["id"]] = {"title_vi": t_vi, "summary_vi": s_vi}

    # Soft cap on cache size
    if len(_trans_cache) > 5000:
        for k in list(_trans_cache.keys())[: len(_trans_cache) - 4000]:
            _trans_cache.pop(k, None)

    return items

File: src/lib/server/translate.ts
import type { NewsItem } from "./rss";

const TRANSLATE_URL = "https://translate.googleapis.com/translate_a/single";

// In-memory cache: item id -> translations. Lives for the duration of the serverless function instance.
const cache = new Map<string, { title_vi: string; summary_vi: string }>();

async function translateOne(text: string): Promise<string> {
  if (!text) return text;
  const truncated = text.slice(0, 5000);
  const params = new URLSearchParams({
    client: "gtx",
    sl: "en",
    tl: "vi",
    dt: "t",
    q: truncated,
  });
  try {
    const r = await fetch(`${TRANSLATE_URL}?${params}`, {
      headers: { "User-Agent": "Mozilla/5.0" },
    });
    if (!r.ok) return text;
    const data = await r.json();
    if (!Array.isArray(data) || !data[0]) return text;
    const parts = (data[0] as unknown[])
      .filter((seg): seg is [string, ...unknown[]] => Array.isArray(seg) && typeof seg[0] === "string")
      .map((seg) => seg[0]);
    return parts.join("").trim() || text;
  } catch {
    return text;
  }
}

export async function translateItems(items: NewsItem[]): Promise<NewsItem[]> {
  const todo: { idx: number; title: string; summary: string }[] = [];

  for (let i = 0; i < items.length; i++) {
    const cached = cache.get(items[i].id);
    if (cached) {
      items[i].title_vi = cached.title_vi;
      items[i].summary_vi = cached.summary_vi;
    } else {
      todo.push({ idx: i, title: items[i].title, summary: items[i].summary });
    }
  }

  if (!todo.length) return items;

  // Translate with limited concurrency (6 at a time)
  const CONCURRENCY = 6;
  for (let i = 0; i < todo.length; i += CONCURRENCY) {
    const batch = todo.slice(i, i + CONCURRENCY);
    await Promise.all(
      batch.map(async ({ idx, title, summary }) => {
        const [title_vi, summary_vi] = await Promise.all([
          translateOne(title),
          summary ? translateOne(summary) : Promise.resolve(""),
        ]);
        items[idx].title_vi = title_vi;
        items[idx].summary_vi = summary_vi;
        cache.set(items[idx].id, { title_vi, summary_vi });
      })
    );
  }

  // Soft cap on cache size
  if (cache.size > 5000) {
    const keys = [...cache.keys()].slice(0, cache.size - 4000);
    for (const k of keys) cache.delete(k);
  }

  return items;
}
